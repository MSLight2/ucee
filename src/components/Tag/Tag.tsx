import React, { useCallback, useState } from 'react';
import classnames from 'classnames';
import './style/index.scss';
import Icon from '../Icon';

export type TagType = 'primary' | 'success' | 'danger' | 'warning'
export enum TAG_TYPE {
  primary = 'primary',
  success = 'success',
  danger = 'danger',
  warning = 'warning'
}

export enum TAG_SIZE {
  medium = 'medium',
  large= 'large'
}

export interface TagProps extends React.HtmlHTMLAttributes<HTMLSpanElement> {
  type?: TagType
  size?: 'medium' | 'large'
  color?: string
  textColor?: string
  plain?: boolean
  round?: boolean
  mark?: boolean
  markDir?: 'left' | 'right'
  closeable?: boolean,
  className?: string
  onClose?: () => void;
}

const Tag: React.FC<TagProps> = (props) => {
  const {
    type = TAG_TYPE.primary,
    size,
    color,
    textColor,
    plain,
    round,
    mark,
    markDir = 'right',
    closeable,
    className,
    children,
    onClose
  } = props;
  const [isCloseable, setIsCloseable] = useState<boolean>(false);
  const baseClass = 'ucee-tag';
  const tagClassName = classnames(
    baseClass,
    {
      [`${baseClass}--primary`]: type === TAG_TYPE.primary,
      [`${baseClass}--success`]: type === TAG_TYPE.success,
      [`${baseClass}--danger`]: type === TAG_TYPE.danger,
      [`${baseClass}--warning`]: type === TAG_TYPE.warning,
      [`${baseClass}--medium`]: size === TAG_SIZE.medium,
      [`${baseClass}--large`]: size === TAG_SIZE.large,
      [`${baseClass}--plain`]: plain,
      [`${baseClass}--round`]: round,
      [`${baseClass}--mark`]: !round && mark,
      [`${baseClass}--mark-dir`]: !round && mark && markDir === 'left',
      [`${baseClass}--close`]: closeable
    },
    className
  );
  const customStyle: any = {};
  if (color) {
    customStyle.backgroundColor = color;
    if (plain) {
      customStyle.backgroundColor = 'white';
      customStyle.color = color;
    }
  }
  if (textColor) {
    customStyle.color = textColor;
  }

  const CloseIcon = useCallback(() => {
    return (
      <>
        {closeable &&
          <Icon name='close' size={12} color="#fff" onClick={() => {
            setIsCloseable(true);
            onClose && typeof onClose === 'function' && onClose();
          }} />
        }
      </>
    )
  }, [closeable, onClose])

  return (
    <>
      {!isCloseable &&
        <span className={tagClassName} style={customStyle}>
          {children}
          <CloseIcon />
        </span>
      }
    </>
  )
}

export default Tag;