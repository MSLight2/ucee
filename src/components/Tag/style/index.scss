@import '../../../styles/var';
@import './mixin.scss';

.ucee-tag{
  position: relative;
  display: inline-block;
  padding: 0 $tag-default-padding;
  color: $white;
  font-size: $tag-default-size;
  line-height: $tag-line-height;
  border-radius: 2px;
  &--primary{
    background-color: $tag-primary-color;
  }
  &--success{
    background-color: $tag-success-color;
  }
  &--danger{
    background-color: $tag-danger-color;
  }
  &--warning{
    background-color: $tag-warning-color;
  }
  &--plain{
    background-color: $white;
    @include tab-plain('ucee-tag');
    &::before{
      content: ' ';
      position: absolute;
      top: 0;
      bottom: 0;
      left: 0;
      right: 0;
      border: 1px solid currentColor;
      border-radius: inherit;
      pointer-events: none;
    }
  }
  &--medium{
    padding: $tag-medium-padding;
  }
  &--large{
    padding: $tag-large-padding;
  }
  &--round{
    border-radius: $border-radius-max;
  }
  &--mark{
    border-radius: 0 $border-radius-max $border-radius-max 0;
    &-dir{
      border-radius: $border-radius-max 0 0 $border-radius-max;
    }
  }
  &--close{
    display: inline-flex;
    align-items: center;
    justify-content: center;
  }
}