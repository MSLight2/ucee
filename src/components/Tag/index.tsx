import Tag from './Tag';

export type { TagType, TagProps } from './Tag';
export { TAG_SIZE, TAG_TYPE } from './Tag';
export default Tag;