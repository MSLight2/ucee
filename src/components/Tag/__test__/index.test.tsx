import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Tag from '../Tag';

describe('RadioGroup', () => {
  test('should render tag with color prop', () => {
    const { container } = render(<Tag color='#d78bde' textColor="#e6160b" />);
    expect(container).toMatchSnapshot();
  });

  test('should render tag with plain prop', () => {
    const { container } = render(<Tag color='#19fad7' plain />);
    expect(container).toMatchSnapshot();
  });

  test('should render tag with mark prop', () => {
    const { container } = render(<Tag mark />);
    expect(container).toMatchSnapshot();
  });

  test('should triggle onClose', () => {
    const onClose = jest.fn();
    const { container } = render(<Tag closeable onClose={onClose} />);
    const closeEle = container.querySelector('.ucee-icon') as Element;

    fireEvent.click(closeEle);
    expect(onClose).toHaveBeenCalled();
    expect(container.querySelector('.ucee-tag')).toBeFalsy();
  });
});