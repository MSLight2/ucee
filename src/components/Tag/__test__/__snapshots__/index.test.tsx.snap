// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup should render tag with color prop 1`] = `
<div>
  <span
    class="ucee-tag ucee-tag--primary"
    style="background-color: rgb(215, 139, 222); color: rgb(230, 22, 11);"
  />
</div>
`;

exports[`RadioGroup should render tag with mark prop 1`] = `
<div>
  <span
    class="ucee-tag ucee-tag--primary ucee-tag--mark"
  />
</div>
`;

exports[`RadioGroup should render tag with plain prop 1`] = `
<div>
  <span
    class="ucee-tag ucee-tag--primary ucee-tag--plain"
    style="background-color: white; color: rgb(25, 250, 215);"
  />
</div>
`;
